standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveAlterHandler.java

```diff
@@ -803,6 +805,24 @@ public Partition alterPartition(RawStore msdb, Warehouse wh, String catName, Str
     return oldParts;
   }
 
+  private void checkTableTypeConversion(Database db, Table oldTbl, Table newTbl)
+          throws InvalidOperationException {
+    // If the given DB is enabled for replication and strict managed is false, then table type cannot be changed.
```

Comment:
what about alter partition set location to a different non hive owned directory? what about add partition with location owned by non hive user ?

Reply:
also wouldnt we have to check this for exchange partition also ?

Reply:
This is not in the scope of this patch. As discussed with @thejasmn today, we cannot disable this add partition as it is normal use case for many customers. Also, he recommended not to honour the migration rule of "convert to external if location owner is not hive". Anyways, we cannot guarantee this in some of the corner cases such as "user manually change location permission etc". So, let's make user to reboot strap such tables after upgrade.
Will send out a mail and shall discuss more on this.